weatherflow2mqtt/weatherflow_mqtt.py

```diff
@@ -166,19 +176,346 @@ async def connect(self) -> None:
+        if (device_class := sensor.device_class) is not None:
+            payload["device_class"] = device_class
+        if (state_class := sensor.state_class) is not None:
+            payload["state_class"] = state_class
+        if (icon := sensor.icon) is not None:
+            payload["icon"] = f"mdi:{icon}"
+        payload["state_topic"] = state_topic
+        payload["value_template"] = f"{{{{ value_json.{sensor.id} }}}}"
+        payload["json_attributes_topic"] = attr_topic
+        payload["device"] = {
+            "identifiers": [f"WeatherFlow_{serial_number}"],
+            "connections": [["mac", serial_number]],
+            "manufacturer": "WeatherFlow",
+            "name": "WeatherFlow2MQTT",
+            "model": f"WeatherFlow Weather Station V{VERSION}",
+            "sw_version": firmware,
+        }
+
+        return payload
+
+    def _handle_observation_event(
+        self, device: WeatherFlowSensorDevice, event: CustomEvent
+    ) -> None:
+        """Handle an observation event."""
+        # Set some class level variables to help with sensors that may not have all data points available:
+        if (val := getattr(device, "solar_radiation", None)) is not None:
+            self.solar_radiation = val.m
+        if (val := getattr(device, "rain_amount_previous_minute", None)) is not None:
+            self.storage["rain_today"] += val.m
+            if val.m > 0:
+                self.storage["rain_duration_today"] += 1
+                self.sql.writeStorage(self.storage)
+
+        state_topic = MQTT_TOPIC_FORMAT.format(
+            DEVICE_SERIAL_FORMAT.format(DOMAIN, device.serial_number),
+            EVENT_OBSERVATION,
+            "state",
+        )
+        data = OrderedDict()
+
+        for sensor in DEVICE_SENSORS:
+            # Skip if this device is missing the attribute
+            if (
+                attr := getattr(device, sensor.device_attr, ATTRIBUTE_MISSING)
+            ) is ATTRIBUTE_MISSING or (
+                sensor.id == "battery_mode" and not isinstance(device, TempestDevice)
+            ):
+                continue
+            try:
+                if isinstance(sensor, SensorDescription):
+                    # TODO: Handle unique data points more elegantly...
+                    if sensor.id == "pressure_trend":
+                        continue
+
+                    # Check for a custom function
+                    if (fn := sensor.custom_fn) is not None:
+                        # TODO: Handle unique data points more elegantly
+                        if sensor.id == "feelslike":
+                            attr = fn(self.cnv, device, self.wind_speed)
+                        elif sensor.id in ("sealevel_pressure", "visibility"):
+                            attr = fn(self.cnv, device, self.elevation)
+                        elif sensor.id == "wbgt":
+                            attr = fn(self.cnv, device, self.solar_radiation)
```

Comment:
You'll notice a few "TODO"s because I haven't figured out the best way to handle these yet while I was working on everything else. I wanted to leave a note in it so that when I see it I can try to continue thinking through it.

Reply:
I have some "real" work for the next hours, so it will take a bit before I get to testing it.

Reply:
We can address the TODOs in another release